cmd/tether/attach.go

```diff
@@ -0,0 +1,369 @@
+func (t *attachServerSSH) run() error {
+	var sConn *ssh.ServerConn
+	var chans <-chan ssh.NewChannel
+	var reqs <-chan *ssh.Request
+	var err error
+
+	// keep waiting for the connection to establish
+	for t.enabled && sConn == nil {
+		// wait for backchannel to establish
+		conn, err := utils.backchannel(context.Background())
+		t.conn = &conn
+
+		// create the SSH server
+		sConn, chans, reqs, err = ssh.NewServerConn(*t.conn, t.config)
+		if err != nil {
+			detail := fmt.Sprintf("failed to establish ssh handshake: %s", err)
+			log.Error(detail)
+			continue
+		}
+
+		defer sConn.Close()
+	}
+	if err != nil {
+		detail := fmt.Sprintf("abandoning attempt to start attach server: %s", err)
+		log.Error(detail)
+		return err
+	}
+
+	// Global requests
+	go t.globalMux(reqs)
+
+	log.Println("ready to service attach requests")
+	// Service the incoming channels
+	for attachchan := range chans {
+		// The only channel type we'll support is attach
+		if attachchan.ChannelType() != attachChannelType {
+			detail := fmt.Sprintf("unknown channel type %s", attachchan.ChannelType())
+			log.Error(detail)
+			attachchan.Reject(ssh.UnknownChannelType, "unknown channel type")
+			continue
+		}
+
+		// check we have a Session matching the requested ID
+		bytes := attachchan.ExtraData()
+		if bytes == nil {
+			detail := "attach channel requires ID in ExtraData"
+			log.Error(detail)
+			attachchan.Reject(ssh.Prohibited, detail)
+			continue
+		}
+
+		sessionid := string(bytes)
+		live, ok := sessions[sessionid]
+
+		reason := ""
+		if !ok || live.cmd == nil {
+			reason = "is unknown"
+		} else if live.cmd.Process == nil {
+			reason = "process has not been launched"
+		} else if live.cmd.Process.Signal(syscall.Signal(0)) != nil {
+			reason = "process has exited"
+		}
+
+		if reason != "" {
+			detail := fmt.Sprintf("attach request: session %s %s", sessionid, reason)
+			log.Error(detail)
+			attachchan.Reject(ssh.Prohibited, detail)
+			continue
+		}
+
+		log.Debugf("accepting incoming channel for %s", sessionid)
+		channel, requests, err := attachchan.Accept()
+		log.Debugf("accepted incoming channel for %s", sessionid)
+		if err != nil {
+			detail := fmt.Sprintf("could not accept channel: %s", err)
+			log.Println(detail)
+			continue
+		}
+
+		// bind the channel to the Session
+		if live.pty == nil {
+			// if it's not a TTY then bind the channel directly to the multiwriter that's already associated with the process
+			dmwStdout, okA := live.cmd.Stdout.(dio.DynamicMultiWriter)
+			dmwStderr, okB := live.cmd.Stderr.(dio.DynamicMultiWriter)
+			dmrStdin, okC := live.cmd.Stdin.(dio.DynamicMultiReader)
+			if !okA || !okB || !okC {
+				detail := fmt.Sprintf("target session IO cannot be duplicated to attach streams: %s", string(bytes))
+				log.Error(detail)
+				attachchan.Reject(ssh.ConnectionFailed, detail)
+				continue
+			}
+
+			log.Debugf("binding reader/writers for channel for %s", sessionid)
```

Comment:
Might want to use the `trace` package here if you're concerned about transaction semantics and just want output of `start` `finish`.